File: 2237-longest-palindrome-by-concatenating-two-letter-words/2237-longest-palindrome-by-concatenating-two-letter-words.java
class Solution {
    public int longestPalindrome(String[] words) {
        Map<String, Integer> map = new HashMap<>();
        int res = 0;
        boolean hasCenter = false;

        for(String word : words){
            String reversed = "" + word.charAt(1) + word.charAt(0);
            if(map.getOrDefault(reversed, 0) > 0){
                res += 4;
                map.put(reversed, map.get(reversed) - 1);
            }else{
                map.put(word, map.getOrDefault(word, 0) + 1);
            }
        }

        for(String key : map.keySet()){
            if(key.charAt(0) == key.charAt(1) && map.get(key) > 0){
                hasCenter = true;
                break;
            }
        }
        if(hasCenter){
            res += 2;
        }
        return res;
    }
}

File: 2237-longest-palindrome-by-concatenating-two-letter-words/2237-longest-palindrome-by-concatenating-two-letter-words.cpp
class Solution {
public:
    int longestPalindrome(vector<string>& words) {
        unordered_map<string, int> mp;

        int res = 0;

        for(string &word : words){
            string reversedWord = word;
            swap(reversedWord[0], reversedWord[1]);

            if(mp[reversedWord] > 0){
                res += 4;
                mp[reversedWord]--;
            }else{
                mp[word]++;
            }
        }
        // Check equal character words, Use only one
        for(auto &it : mp){
            string word = it.first;
            int count = it.second;

            if(word[0] == word[1] && count > 0){
                res += 2;
                break;
            }
        }
        return res;
    }
};